### ISSUE TITLE
Timeouts in lamda code leading to weird issue

### ISSUE THREAD

**Issue by cosmok** [brefphp/bref#639]
Hi,

Thanks to the team for your amazing work!

I have used bref in a few projects and I am happy with it.

In one of the new projects that I am involved in, I received reports about timeouts persisting across requests. In other words, a timeout in Request A is sometimes leading to an 'internal server error' message response for another subsequent Request B.

Following is a setup that can reproduce this issue:
serverless.yml
```
service: demo-timeout

provider:
    name: aws
    stage: staging
    logs:
      restApi: true
    region: ap-southeast-2
    runtime: provided
    vpc: 'vpc-id'
    endpointType: 'regional' 

plugins:
    - ./vendor/bref/bref

functions:
    api:
        handler: index.php
        memorySize: 1024
        description: ''
        timeout: 5 # in seconds (API Gateway has a timeout of 29 seconds)
        layers:
            - ${bref:layer.php-72-fpm}
        events:
           -   http: 'GET /'
``` 

index.php
```
<?php

require __DIR__ . '/vendor/autoload.php';

/**
 * a simple script that concatenates
 * two passed params 'a' and 'b'
 * and optionally simulates a socket timeout
 * based off a passed param 'timeout'
 */
$a = ($_GET['a']) ?? 'Hello';
$b = ($_GET['b']) ?? 'World!';

error_log('inside function');
if (isset($_GET['timeout']) && ctype_digit($_GET['timeout'])) {
    $timeout = (int) $_GET['timeout'];
    error_log("timeout for: {$timeout} seconds");
    //non-routable ip
    fsockopen('10.255.255.1', 443, $error_no, $error_string, $timeout);
}
$message = "{$a} {$b}";
header("Content-Type: application/json");
echo json_encode([
    'message' => $message,
]);
```
composer.json
```
{
    "require": {
        "bref/bref": "dev-master#404c69f2f"
    }
}
```
Using latest bref - tried the latest master just in case as the current release has the issue.
No other packages or extensions involved.

Following siege command will reveal the issue:
```
siege -d1 -c1 -v  -t30s -p -f urls.txt 
[error] CONFIG conflict: selected time and repetition based testing
defaulting to time-based testing: 30 seconds
{"message": "Internal server error"}
{"message":"Achyuth Kesavan"}
{"message": "Internal server error"}
{"message": "Internal server error"}
{"message": "Internal server error"}
{"message": "Internal server error"}
{"message": "Internal server error"}
{"message":"Achyuth Kesavan"}
{"message": "Internal server error"}
{"message": "Internal server error"}
{"message": "Internal server error"}
{"message": "Internal server error"}
{"message": "Internal server error"}
{"message": "Internal server error"}
{"message": "Internal server error"}
{"message":"Achyuth Kesavan"}

{	"transactions":			           3,
	"availability":			       18.75,
	"elapsed_time":			       29.15,
	"data_transferred":		        0.00,
	"response_time":		        4.66,
	"transaction_rate":		        0.10,
	"throughput":			        0.00,
	"concurrency":			        0.48,
	"successful_transactions":	           3,
	"failed_transactions":		          13,
	"longest_transaction":		        5.48,
	"shortest_transaction":		        0.15
}
```
urls.txt
```
https://{{API_GW_ID}}.execute-api.ap-southeast-2.amazonaws.com/staging?a=Achyuth&b=KesavanT&timeout=5
https://{{API_GW_ID}}.execute-api.ap-southeast-2.amazonaws.com/staging?a=Achyuth&b=Kesavan
```
Cloudwatch logs:

```
START RequestId: db127b92-13b8-43ad-8e21-2b826ef58524 Version: $LATEST
PHP-FPM seems to be running already, this might be because Lambda stopped the bootstrap process but didn't leave us an opportunity to stop PHP-FPM. Stopping PHP-FPM now to restart from a blank slate.
RequestId: db127b92-13b8-43ad-8e21-2b826ef58524 Error: Runtime exited with error: signal: terminated
Runtime.ExitError
END RequestId: db127b92-13b8-43ad-8e21-2b826ef58524
REPORT RequestId: db127b92-13b8-43ad-8e21-2b826ef58524	Duration: 268.11 ms	Billed Duration: 300 ms	Memory Size: 1024 MB	Max Memory Used: 33 MB	
START RequestId: 82f8c759-0a78-457e-a34f-3e61e216352b Version: $LATEST
PHP-FPM seems to be running already, this might be because Lambda stopped the bootstrap process but didn't leave us an opportunity to stop PHP-FPM. Stopping PHP-FPM now to restart from a blank slate.
END RequestId: 82f8c759-0a78-457e-a34f-3e61e216352b
REPORT RequestId: 82f8c759-0a78-457e-a34f-3e61e216352b	Duration: 48.22 ms	Billed Duration: 100 ms	Memory Size: 1024 MB	Max Memory Used: 9 MB	
RequestId: 82f8c759-0a78-457e-a34f-3e61e216352b Error: Runtime exited with error: signal: terminated
Runtime.ExitError
```
The reason I believe that it is a timeout related issue is that, if I modify the first URL in the file above and set the timeout to 2 (which is less than the set lambda timeout), I see:

```
 siege -d1 -c1 -v  -t30s -p -f urls.txt 
[error] CONFIG conflict: selected time and repetition based testing
defaulting to time-based testing: 30 seconds
{"message":"Achyuth KesavanT"}
{"message":"Achyuth Kesavan"}
{"message":"Achyuth KesavanT"}
{"message":"Achyuth Kesavan"}
{"message":"Achyuth KesavanT"}
{"message":"Achyuth Kesavan"}
{"message":"Achyuth KesavanT"}
{"message":"Achyuth Kesavan"}
{"message":"Achyuth KesavanT"}
{"message":"Achyuth Kesavan"}
{"message":"Achyuth KesavanT"}
{"message":"Achyuth Kesavan"}
{"message":"Achyuth KesavanT"}
{"message":"Achyuth Kesavan"}
{"message":"Achyuth KesavanT"}
{"message":"Achyuth Kesavan"}
{"message":"Achyuth KesavanT"}
{"message":"Achyuth Kesavan"}
{"message":"Achyuth KesavanT"}
{"message":"Achyuth Kesavan"}

{	"transactions":			          20,
	"availability":			      100.00,
	"elapsed_time":			       29.83,
	"data_transferred":		        0.00,
	"response_time":		        1.23,
	"transaction_rate":		        0.67,
	"throughput":			        0.00,
	"concurrency":			        0.82,
	"successful_transactions":	          20,
	"failed_transactions":		           0,
	"longest_transaction":		        3.41,
	"shortest_transaction":		        0.11
}
```
siege is picking up the two urls one by one and the response is as expected.

I also tried this with a nodejs setup and it wasn't exhibiting this issue.

I can see #272 that should have fixed this, but, perhaps there is some race condition that is causing this.

Sorry for the long report.

Appreciate any help!
:)

**Comment by mnapoli** [comment]
Thank you for that detailed report!

If I summarize, what you suspect is:

- request A has a *Lambda timeout*, leaving PHP-FPM in a "still running" state
- request B arrives, detects that PHP-FPM is not finished, tries to kill it but somehow that doesn't work and we get an internal error

What you could do to continue debugging is editing that file in your `vendor/bref/bref` directory:

https://github.com/brefphp/bref/blob/404c69f2f293c2fe0841835cc8addac72575791d/src/Event/Http/FpmHandler.php#L223-L262

For example, you could add `echo` statements all over the place and see what gets executed and what doesn't.

If we look at the logs:

```
START RequestId: db127b92-13b8-43ad-8e21-2b826ef58524 Version: $LATEST
PHP-FPM seems to be running already, this might be because Lambda stopped the bootstrap process but didn't leave us an opportunity to stop PHP-FPM. Stopping PHP-FPM now to restart from a blank slate.
RequestId: db127b92-13b8-43ad-8e21-2b826ef58524 Error: Runtime exited with error: signal: terminated
Runtime.ExitError
END RequestId: db127b92-13b8-43ad-8e21-2b826ef58524
REPORT RequestId: db127b92-13b8-43ad-8e21-2b826ef58524	Duration: 268.11 ms	Billed Duration: 300 ms	Memory Size: 1024 MB	Max Memory Used: 33 MB	
```

we can guess that the PHP process crashed somewhere after this line:

https://github.com/brefphp/bref/blob/404c69f2f293c2fe0841835cc8addac72575791d/src/Event/Http/FpmHandler.php#L231

Maybe the `posix_kill` call failed? That is why adding `echo` statements to debug could help to see where the:

> Error: Runtime exited with error: signal: terminated

happens.

**Comment by cosmok** [comment]
Thank for the pointer! 
I will try those and report back.

**Comment by cosmok** [comment]
Hi @mnapoli ,
Sorry, I didn't have the time to look into this until today.
As per your suggestion, I started adding debug statements and I found that there was nothing in the logs after posix_kill call.
It turns out that it killed/terminated itself. See my PR which solves this issue :)

I would like to also get someone to check a few things:
* __destruct() calls stop() which can throw an Exception
* $this->fpm->setTimeout(null), perhaps the timeout should be closer to the lambda timeout
* $this->fpm->stop(2), why 2?
